conductor-client/src/main/kotlin/com/openlattice/postgres/external/ExternalDatabasePermissioner.kt

```diff
@@ -0,0 +1,515 @@
+            }
+        }
+        updateAssemblyPermissions(action, assemblyAcls, aclKeyToTableCols.toCompletableFuture().get())
+    }
+
+    private fun mapAclsToExternalVsAssemblyColumnAcls(acls: List<Acl>): Map<SecurableObjectType, List<Acl>>{
+        val aclKeyIndex = acls.aclKeysAsSet { it.aclKey.index }.toTypedArray()
+
+        val allOrgExternalDBAclKeys: Set<AclKey> = securableObjectTypes.keySet(
+                getAclKeysOfObjectTypePredicate( aclKeyIndex, SecurableObjectType.OrganizationExternalDatabaseColumn)
+        )
+        val allOrgExternalDbAcls = acls.filter { allOrgExternalDBAclKeys.contains(it.aclKey) }
+
+        val allAssemblyAclKeys: Set<AclKey> = securableObjectTypes.keySet(
+                getAclKeysOfObjectTypePredicate( aclKeyIndex, SecurableObjectType.PropertyTypeInEntitySet)
+        )
+        val allAssemblyAcls = acls.filter { allAssemblyAclKeys.contains(it.aclKey) }
+
+        return mapOf(
+                SecurableObjectType.OrganizationExternalDatabaseColumn to allOrgExternalDbAcls,
+                SecurableObjectType.PropertyTypeInEntitySet to allAssemblyAcls
+        )
+    }
+
+    private fun getAclKeysOfObjectTypePredicate(
+            aclKeysIndexForm: Array<String>,
+            objectType: SecurableObjectType
+    ): Predicate<AclKey, SecurableObjectType> {
+        return Predicates.and<AclKey, SecurableObjectType>(
+                Predicates.`in`<Any, Any>(SecurableObjectTypeMapstore.ACL_KEY_INDEX, *aclKeysIndexForm),
+                Predicates.equal<Any, Any>(SecurableObjectTypeMapstore.SECURABLE_OBJECT_TYPE_INDEX, objectType )
+        )
+    }
+
+    private fun <T> Collection<Acl>.aclKeysAsSet( extraction: ( Acl ) -> T ): Set<T> {
+        return this.mapTo( HashSet(this.size) ) { extraction(it) }
+    }
+
+    override fun addPrincipalToPrincipals(sourcePrincipalAclKey: AclKey, targetPrincipalAclKeys: Set<AclKey>) {
+        if ( targetPrincipalAclKeys.isEmpty()) {
+            return
+        }
+        val grantTargets = dbCredentialService.getDbUsernames(targetPrincipalAclKeys)
+
+        val sourceRole = dbCredentialService.getDbUsername(sourcePrincipalAclKey)
+
+        logger.debug("attempting to grant $sourceRole to $grantTargets")
+        atlas.connection.use { connection ->
+            connection.createStatement().use { stmt ->
+                stmt.execute(grantRoleToRole(sourceRole, grantTargets))
+            }
+        }
+    }
+
+    override fun removePrincipalsFromPrincipals(principalsToRemove: Set<AclKey>, fromPrincipals: Set<AclKey>) {
+        val removeFrom = principalsToPostgresRoleNames(fromPrincipals)
+        atlas.connection.use { connection ->
+            connection.createStatement().use { stmt ->
+                principalsToRemove.forEach { aclKey ->
+                    val roleName = PostgresRoles.buildPostgresRoleName(principalsMapManager.lookupRole(aclKey).id)
```

Comment:
these role names should be looked up using dbCreds -- not guaranteed that they're roles, and also these individual calls (vs bulk up front) will slow this down a lot (depending how many principals there are)

Reply:
Also probably good to move this out of the open connection as well